{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Meth_Comp CLI usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Activate virtual environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T22:08:40.854666Z",
     "start_time": "2020-05-18T22:08:40.601Z"
    },
    "init_cell": true
   },
   "outputs": [],
   "source": [
    "# Activate Conda env \n",
    "conda activate pycoMeth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T22:08:43.784683Z",
     "start_time": "2020-05-18T22:08:42.436679Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: pycoMeth Comp_Report [-h] -i METHCOMP_FN -g GFF3_FN -f REF_FASTA_FN\n",
      "                            [-o OUTDIR] [-n N_TOP] [-d MAX_TSS_DISTANCE]\n",
      "                            [--pvalue_threshold PVALUE_THRESHOLD]\n",
      "                            [--min_diff_llr MIN_DIFF_LLR]\n",
      "                            [--n_len_bin N_LEN_BIN] [-v] [-q] [-p]\n",
      "\n",
      "Generate an HTML report of significantly differentially methylated CpG\n",
      "intervals from `Meth_Comp` text output. Significant intervals are annotated\n",
      "with their closest transcript TSS.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input/Output options:\n",
      "  -i METHCOMP_FN, --methcomp_fn METHCOMP_FN\n",
      "                        Input tsv file generated by Meth_comp (can be\n",
      "                        gzipped). At the moment only data binned by intervals\n",
      "                        with Interval_Aggregate are supported. (required)\n",
      "                        [str]\n",
      "  -g GFF3_FN, --gff3_fn GFF3_FN\n",
      "                        Path to an **ensembl GFF3** file containing genomic\n",
      "                        annotations. Only the transcripts details are\n",
      "                        extracted. (required) [str]\n",
      "  -f REF_FASTA_FN, --ref_fasta_fn REF_FASTA_FN\n",
      "                        Reference file used for alignment in Fasta format\n",
      "                        (ideally already indexed with samtools faidx)\n",
      "                        (required) [str]\n",
      "  -o OUTDIR, --outdir OUTDIR\n",
      "                        Directory where to output HTML reports, By default\n",
      "                        current directory (default: None) [str]\n",
      "\n",
      "Misc options:\n",
      "  -n N_TOP, --n_top N_TOP\n",
      "                        Number of top interval candidates for which to\n",
      "                        generate an interval report. If there are not enough\n",
      "                        significant candidates this is automatically scaled\n",
      "                        down. (default: 100) [int]\n",
      "  -d MAX_TSS_DISTANCE, --max_tss_distance MAX_TSS_DISTANCE\n",
      "                        Maximal distance to transcription stat site to find\n",
      "                        transcripts close to interval candidates (default:\n",
      "                        500000) [int]\n",
      "  --pvalue_threshold PVALUE_THRESHOLD\n",
      "                        pValue cutoff for top interval candidates (default:\n",
      "                        0.01) [float]\n",
      "  --min_diff_llr MIN_DIFF_LLR\n",
      "                        Minimal llr boundary for negative and positive median\n",
      "                        llr. 1 is recommanded for vizualization purposes.\n",
      "                        (default: 1) [float]\n",
      "  --n_len_bin N_LEN_BIN\n",
      "                        Number of genomic intervals for the longest chromosome\n",
      "                        of the ideogram figure (default: 500) [int]\n",
      "\n",
      "Verbosity options:\n",
      "  -v, --verbose         Increase verbosity\n",
      "  -q, --quiet           Reduce verbosity\n",
      "  -p, --progress        Display a progress bar\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Comp_Report --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example with a single significant result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T22:08:54.297213Z",
     "start_time": "2020-05-18T22:08:51.936554Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: Options summary\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tPackage name: pycoMeth\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tPackage version: 0.4.3\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tTimestamp: 2020-05-18 23:08:53.283886\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tmethcomp_fn: ./data/Yeast_CGI_meth_comp.tsv.gz\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tgff3_fn: ./data/yeast.gff3\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tref_fasta_fn: ./data/yeast.fa\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \toutdir: yeast_html\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tn_top: 100\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tmax_tss_distance: 500000\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tpvalue_threshold: 0.05\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tmin_diff_llr: 1\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tn_len_bin: 500\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tverbose: True\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tquiet: False\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tprogress: False\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tkwargs\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \t\tsubcommands: Comp_Report\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \t\tfunc: <function Comp_Report at 0x7f5b2f140950>\u001b[0m\n",
      "\u001b[01;34m## Loading and preparing data ##\u001b[0m\n",
      "\u001b[32m\tLoading Methcomp data from TSV file\u001b[0m\n",
      "\u001b[32m\tLoading transcripts info from GFF file\u001b[0m\n",
      "\u001b[32m\tLoading chromosome info from reference FASTA file\u001b[0m\n",
      "\u001b[32m\tNumber of significant intervals found (adjusted pvalue<0.05): 1\u001b[0m\n",
      "\u001b[01;31mERROR: Low number of significant sites. The summary report will likely contains errors\u001b[0m\n",
      "\u001b[01;31mERROR: Number of significant intervals lower than number of top candidates to plot\u001b[0m\n",
      "\u001b[32m\tGenerating file names for top candidates reports\u001b[0m\n",
      "\u001b[32m\tComputing source md5\u001b[0m\n",
      "\u001b[01;34m## Parsing methcomp data ##\u001b[0m\n",
      "\u001b[32m\tIterating over significant intervals and generating top candidates reports\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: Ploting top candidates: ('V', 65380, 65612)\u001b[0m\n",
      "\u001b[32m\tGenerating summary report\u001b[0m\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Comp_Report \\\n",
    "    -i \"./data/Yeast_CGI_meth_comp.tsv.gz\" \\\n",
    "    -g \"./data/yeast.gff3\" \\\n",
    "    -f \"./data/yeast.fa\" \\\n",
    "    -o \"yeast_html\" \\\n",
    "    --pvalue_threshold 0.05 \\\n",
    "    --verbose"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with large dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T22:09:34.106897Z",
     "start_time": "2020-05-18T22:09:12.429914Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Loading and preparing data ##\u001b[0m\n",
      "\u001b[32m\tLoading Methcomp data from TSV file\u001b[0m\n",
      "\u001b[32m\tLoading transcripts info from GFF file\u001b[0m\n",
      "\u001b[32m\tLoading chromosome info from reference FASTA file\u001b[0m\n",
      "\u001b[32m\tNumber of significant intervals found (adjusted pvalue<0.01): 3532\u001b[0m\n",
      "\u001b[32m\tGenerating file names for top candidates reports\u001b[0m\n",
      "\u001b[32m\tComputing source md5\u001b[0m\n",
      "\u001b[01;34m## Parsing methcomp data ##\u001b[0m\n",
      "\u001b[32m\tIterating over significant intervals and generating top candidates reports\u001b[0m\n",
      "\tProgress: 100%|████████████████████| 3.53k/3.53k [00:16<00:00, 210 intervals/s]\n",
      "\u001b[32m\tGenerating summary report\u001b[0m\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Comp_Report \\\n",
    "    -i \"./data/Medaka_CGI_meth_comp.tsv.gz\" \\\n",
    "    -g \"./data/medaka.gff3\" \\\n",
    "    -f \"./data/medaka.fa\" \\\n",
    "    -o \"medaka_html\" \\\n",
    "    --n_top 100 \\\n",
    "    --progress"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
