![](pictures/pycoMeth_long.png)

[![GitHub license](https://img.shields.io/github/license/a-slide/pycoMeth.svg)](https://github.com/a-slide/pycoMeth/blob/master/LICENSE)
[![Language](https://img.shields.io/badge/Language-Python3-yellow.svg)](https://www.python.org/)
[![PyPI version](https://badge.fury.io/py/pycoMeth.svg)](https://badge.fury.io/py/pycoMeth)
[![PyPI downloads](https://pepy.tech/badge/pycoMeth)](https://pepy.tech/project/pycoMeth)
[![DOI](https://zenodo.org/badge/211195001.svg)](https://zenodo.org/badge/latestdoi/211195001)
[![Build Status](https://travis-ci.com/a-slide/pycoMeth.svg?branch=master)](https://travis-ci.com/a-slide/pycoMeth)

---

# Welcome to pycoMeth documentation

---

**DNA methylation analysis downstream to Nanopolish for Oxford Nanopore DNA sequencing datasets**

`pycoMeth` can be used for further analyses starting from the output files generated by [`Nanopolish call-methylation`](https://github.com/jts/nanopolish). The package contains a suite of tools to **find CpG islands** calculate the **methylation probability at CpG dinucleotide or CpG island resolution** across the entire genome and to perform a **differential methylation analysis** across multiple samples. Finally an **interactive HTML report** can be generated to visualize significant intervals.

The package contains the following modules:

* [CGI_Finder](CGI_Finder/usage): Find CpG islands in a fasta file by using a sliding window and merging overlapping windows satisfying the CpG island definition.
* [CpG_Aggregate](CpG_Aggregate/usage): Aggregate the output of `nanopolish call-methylation` at genomic position level.
* [Interval_Aggregate](Interval_Aggregate/usage): Bin the output of `pycoMeth CpG_Aggregate` in genomic intervals, using either an annotation file or a sliding window.
* [Meth_Comp](Meth_Comp/usage): Compare Methylation level between several samples from either `pycoMeth CpG_Aggregate` or `pycoMeth Interval_Aggregate`.
* [Comp_Report](Comp_Report/usage): Generate an interactive HTML report for top significant differentially methylated intervals found with `pycoMeth Meth_Comp`

`pycoMeth` generates extensive tabulated reports and BED files which can be loaded in a genome browser. However, this is not a visualisation tool *per se*. An excellent complementary tool to visualise and explore methylation status for specific loci is [`Methplotlib`](https://github.com/wdecoster/methplotlib) developed by [Wouter de coster](https://twitter.com/wouter_decoster).


Please be aware that `pycoMeth` is a research package that is still under development.
The API, command line interface, and implementation might change without retro-compatibility.

---

### pycoMeth workflow

![Workflow](pictures/pycoMeth_package.png)

### pycoMeth example output IGV rendering

![IGV](pictures/pycoMeth_all.png)

### pycoMeth example HTML report

[Example HTML report](https://a-slide.github.io/pycoMeth/Comp_Report/medaka_html/pycoMeth_summary_report.html)

![HTML](pictures/pycoMeth_HTML.gif)

---

### Citing

The repository is archived at Zenodo. If you use `pycoMeth` please cite as follow:

Adrien Leger. (2020, January 28). a-slide/pycoMeth. Zenodo. https://doi.org/10.5281/zenodo.3629254

### Authors

* Adrien Leger - aleg {at} ebi.ac.uk
