{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Meth_Comp API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T21:52:36.822758Z",
     "start_time": "2020-05-18T21:52:36.817209Z"
    },
    "init_cell": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Import main module \n",
    "from pycoMeth.Meth_Comp import Meth_Comp\n",
    "\n",
    "# optionally inport jupyter helper functions\n",
    "from pycoMeth.common import head, jhelp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T21:52:49.066551Z",
     "start_time": "2020-05-18T21:52:49.044251Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**Meth_Comp** (aggregate_fn_list, ref_fasta_fn, output_tsv_fn, output_bed_fn, max_missing, min_diff_llr, sample_id_list, pvalue_adj_method, pvalue_threshold, only_tested_sites, verbose, quiet, progress, kwargs)\n",
       "\n",
       "Compare methylation values for each CpG positions or intervals between n samples and perform a statistical test to evaluate if the positions are significantly different. For 2 samples a Mann_Withney test is performed otherwise multiples samples are compared with a Kruskal Wallis test. pValues are adjusted for multiple tests using the Benjamini & Hochberg procedure for controlling the false discovery rate.\n",
       "\n",
       "---\n",
       "\n",
       "* **aggregate_fn_list** (required) [list(str)]\n",
       "\n",
       "A list of output tsv files corresponding to several samples to compare generated by either CpG_Aggregate or Interval_Aggregate. (can be gzipped)\n",
       "\n",
       "* **ref_fasta_fn** (required) [str]\n",
       "\n",
       "Reference file used for alignment in Fasta format (ideally already indexed with samtools faidx)\n",
       "\n",
       "* **output_tsv_fn** (default: None) [str]\n",
       "\n",
       "Path to write an more extensive result report in TSV format (At least 1 output file is required) (can be gzipped)\n",
       "\n",
       "* **output_bed_fn** (default: None) [str]\n",
       "\n",
       "Path to write a summary result file in BED format (At least 1 output file is required) (can be gzipped)\n",
       "\n",
       "* **max_missing** (default: 0) [int]\n",
       "\n",
       "Max number of missing samples to perform the test\n",
       "\n",
       "* **min_diff_llr** (default: 2) [float]\n",
       "\n",
       "Minimal llr boundary for negative and positive median llr. The test if only performed if at least one sample has a median llr above (methylated) and 1 sample has a median llr below (unmethylated)\n",
       "\n",
       "* **sample_id_list** (default: None) [list(str)]\n",
       "\n",
       "list of sample ids to annotate results in tsv file\n",
       "\n",
       "* **pvalue_adj_method** (default: fdr_bh) [str]\n",
       "\n",
       "Method to use for pValue multiple test adjustment\n",
       "\n",
       "* **pvalue_threshold** (default: 0.01) [float]\n",
       "\n",
       "Alpha parameter (family-wise error rate) for pValue adjustment\n",
       "\n",
       "* **only_tested_sites** (default: False) [bool]\n",
       "\n",
       "Do not include sites that were not tested because of insufficient samples or effect size in the report\n",
       "\n",
       "* **verbose** (default: False) [bool]\n",
       "\n",
       "* **quiet** (default: False) [bool]\n",
       "\n",
       "* **progress** (default: False) [bool]\n",
       "\n",
       "* **kwargs**\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(Meth_Comp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with CpG Aggregate output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T21:54:38.885949Z",
     "start_time": "2020-05-18T21:54:31.578797Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing files ##\u001b[0m\n",
      "\u001b[32m\tReading input files header and checking consistancy between headers\u001b[0m\n",
      "\u001b[32m\tStarting asynchronous file parsing\u001b[0m\n",
      "\tProgress: 37.1M bytes [00:07, 5.12M bytes/s]                       \n",
      "\u001b[32m\tAdjust pvalues\u001b[0m\n",
      "\u001b[32m\tWriting output file\u001b[0m\n",
      "\tProgress: 100%|██████████| 344/344 [00:00<00:00, 18.8k sites/s]\n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tInsufficient samples: 211,400\u001b[0m\n",
      "\u001b[32m\t\tInsufficient effect size: 33,071\u001b[0m\n",
      "\u001b[32m\t\tValid: 344\u001b[0m\n",
      "\u001b[32m\t\tNon-significant pvalue: 344\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome start  end    n_samples pvalue              adj_pvalue          neg_med pos_med ambiguous_med labels                med_llr_list                raw_llr_list                                       comment                \n",
      "I          118799 118800 4         0.29978058859571194 0.4279805197751711  3       1       0             [\"S1\",\"S2\",\"S3\",\"S4\"] [2.005,-2.645,-6.64,-5.51]  [[-1.82,5.83],[-0.09,-5.2],[-2.67,-10.61],[-6.7...Non-significant pvalue \n",
      "I          141415 141416 4         0.11800901597381579 0.3742064977444858  1       1       2             [\"S1\",\"S2\",\"S3\",\"S4\"] [-4.175,0.135,3.02,-0.02]   [[-1.33,-4.52,-3.83,-4.67],[-1.42,1.69],[6.62,-...Non-significant pvalue \n",
      "I          151819 151820 4         0.1315327225513005  0.3742064977444858  1       1       2             [\"S1\",\"S2\",\"S3\",\"S4\"] [-2.98,-1.95,2.93,-1.48]    [[-3.68,-4.4,0.54,-2.28],[-3.45,-1.95,-1.21],[4...Non-significant pvalue \n",
      "I          167230 167237 4         0.10985439586067039 0.3742064977444858  3       1       0             [\"S1\",\"S2\",\"S3\",\"S4\"] [-6.305,-6.94,3.6,-4.17]    [[-5.06,-7.55],[-6.94,-7.26,-1.12],[3.6,5.68,0....Non-significant pvalue \n",
      "II         64407  64408  4         0.16469672531155308 0.3742064977444858  1       1       2             [\"S1\",\"S2\",\"S3\",\"S4\"] [-1.06,-0.19,-3.68,2.93]    [[-1.06,-0.81,-2.26],[1.65,-2.03],[-3.85,-2.85,...Non-significant pvalue \n",
      "II         102416 102417 4         0.06267205003507882 0.3742064977444858  2       1       1             [\"S1\",\"S2\",\"S3\",\"S4\"] [0.715,-3.05,2.17,-5.535]   [[-0.21,1.64],[-3.3,-2.8],[2.24,2.1],[-4.52,-2....Non-significant pvalue \n",
      "II         108834 108835 4         0.1633613146761696  0.3742064977444858  3       1       0             [\"S1\",\"S2\",\"S3\",\"S4\"] [2.95,-5.16,-3.53,-4.53]    [[4.7,1.2],[-5.89,-4.43],[-6.55,-0.25,-3.53],[-...Non-significant pvalue \n",
      "II         110457 110458 4         0.04436795794391088 0.3742064977444858  2       1       1             [\"S1\",\"S2\",\"S3\",\"S4\"] [-1.405,2.69,-3.285,-7.045] [[-0.97,-1.34,-1.47,-3.97],[-0.84,6.22],[-7.1,0...Non-significant pvalue \n",
      "II         112989 112990 4         0.3618050274975317  0.46268003516412975 1       1       2             [\"S1\",\"S2\",\"S3\",\"S4\"] [-1.38,2.1,-1.94,-2.93]     [[-1.38,4.34,-5.48,-5.45,-0.93],[3.4,0.8],[-1.3...Non-significant pvalue \n",
      "\n",
      "track name=meth_comp itemRgb=On\n",
      "I\t118799\t118800\t.\t0\t.\t118799\t118800\t230,230,230\n",
      "I\t141415\t141416\t.\t0\t.\t141415\t141416\t230,230,230\n",
      "I\t151819\t151820\t.\t0\t.\t151819\t151820\t230,230,230\n",
      "I\t167230\t167237\t.\t0\t.\t167230\t167237\t230,230,230\n",
      "II\t64407\t64408\t.\t0\t.\t64407\t64408\t230,230,230\n",
      "II\t102416\t102417\t.\t0\t.\t102416\t102417\t230,230,230\n",
      "II\t108834\t108835\t.\t0\t.\t108834\t108835\t230,230,230\n",
      "II\t110457\t110458\t.\t0\t.\t110457\t110458\t230,230,230\n",
      "II\t112989\t112990\t.\t0\t.\t112989\t112990\t230,230,230\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ff = Meth_Comp (\n",
    "    aggregate_fn_list=[\n",
    "        \"./data/Yeast_CpG_1.tsv.gz\", \n",
    "        \"./data/Yeast_CpG_2.tsv.gz\", \n",
    "        \"./data/Yeast_CpG_3.tsv.gz\", \n",
    "        \"./data/Yeast_CpG_4.tsv.gz\"],\n",
    "    ref_fasta_fn=\"./data/yeast.fa\",\n",
    "    output_bed_fn=\"./results/Yeast_CpG_meth_comp.bed\",\n",
    "    output_tsv_fn=\"./results/Yeast_CpG_meth_comp.tsv.gz\",\n",
    "    sample_id_list=[\"S1\",\"S2\",\"S3\",\"S4\"],\n",
    "    only_tested_sites=True,\n",
    "    progress=True)\n",
    "\n",
    "head(\"./results/Yeast_CpG_meth_comp.tsv.gz\")\n",
    "head(\"./results/Yeast_CpG_meth_comp.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with Interval Aggregate output with a single significant result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T21:54:21.698740Z",
     "start_time": "2020-05-18T21:54:21.492291Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing files ##\u001b[0m\n",
      "\u001b[32m\tReading input files header and checking consistancy between headers\u001b[0m\n",
      "\u001b[32m\tStarting asynchronous file parsing\u001b[0m\n",
      "\tProgress: 776k bytes [00:00, 6.48M bytes/s]              \n",
      "\u001b[32m\tAdjust pvalues\u001b[0m\n",
      "\u001b[32m\tWriting output file\u001b[0m\n",
      "\tProgress: 100%|██████████| 1.86k/1.86k [00:00<00:00, 36.0k sites/s]\n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tInsufficient effect size: 938\u001b[0m\n",
      "\u001b[32m\t\tInsufficient samples: 921\u001b[0m\n",
      "\u001b[32m\t\tValid: 4\u001b[0m\n",
      "\u001b[32m\t\tNon-significant pvalue: 4\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome start end   n_samples pvalue adj_pvalue neg_med pos_med ambiguous_med unique_cpg_pos labels           med_llr_list raw_llr_list raw_pos_list comment                  \n",
      "I          17    333   2         nan    nan        2       0       0             0              [\"S1\",\"S4\"]      []           []           []           Insufficient samples     \n",
      "I          1804  2170  3         nan    nan        3       0       0             0              [\"S1\",\"S2\",\"S4\"] []           []           []           Insufficient effect size \n",
      "I          31835 32949 1         nan    nan        1       0       0             0              [\"S1\"]           []           []           []           Insufficient samples     \n",
      "I          33497 34371 2         nan    nan        2       0       0             0              [\"S1\",\"S2\"]      []           []           []           Insufficient samples     \n",
      "I          38163 38471 2         nan    nan        2       0       0             0              [\"S1\",\"S2\"]      []           []           []           Insufficient samples     \n",
      "I          44294 44565 2         nan    nan        2       0       0             0              [\"S1\",\"S2\"]      []           []           []           Insufficient samples     \n",
      "I          44730 44988 2         nan    nan        2       0       0             0              [\"S1\",\"S2\"]      []           []           []           Insufficient samples     \n",
      "I          45308 45526 2         nan    nan        2       0       0             0              [\"S1\",\"S2\"]      []           []           []           Insufficient samples     \n",
      "I          46201 46642 2         nan    nan        2       0       0             0              [\"S1\",\"S2\"]      []           []           []           Insufficient samples     \n",
      "\n",
      "track name=meth_comp itemRgb=On\n",
      "I\t17\t333\t.\t0\t.\t17\t333\t230,230,230\n",
      "I\t1804\t2170\t.\t0\t.\t1804\t2170\t230,230,230\n",
      "I\t31835\t32949\t.\t0\t.\t31835\t32949\t230,230,230\n",
      "I\t33497\t34371\t.\t0\t.\t33497\t34371\t230,230,230\n",
      "I\t38163\t38471\t.\t0\t.\t38163\t38471\t230,230,230\n",
      "I\t44294\t44565\t.\t0\t.\t44294\t44565\t230,230,230\n",
      "I\t44730\t44988\t.\t0\t.\t44730\t44988\t230,230,230\n",
      "I\t45308\t45526\t.\t0\t.\t45308\t45526\t230,230,230\n",
      "I\t46201\t46642\t.\t0\t.\t46201\t46642\t230,230,230\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ff = Meth_Comp (\n",
    "    aggregate_fn_list=[\n",
    "        \"./data/Yeast_CGI_1.tsv.gz\", \n",
    "        \"./data/Yeast_CGI_2.tsv.gz\", \n",
    "        \"./data/Yeast_CGI_3.tsv.gz\", \n",
    "        \"./data/Yeast_CGI_4.tsv.gz\"],\n",
    "    ref_fasta_fn=\"./data/yeast.fa\",\n",
    "    output_bed_fn=\"./results/Yeast_CGI_meth_comp.bed\",\n",
    "    output_tsv_fn=\"./results/Yeast_CGI_meth_comp.tsv.gz\",\n",
    "    sample_id_list=[\"S1\",\"S2\",\"S3\",\"S4\"],\n",
    "    max_missing = 1,\n",
    "    min_diff_llr = 0,\n",
    "    only_tested_sites=False,\n",
    "    progress=True)\n",
    "\n",
    "head(\"./results/Yeast_CGI_meth_comp.tsv.gz\")\n",
    "head(\"./results/Yeast_CGI_meth_comp.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with Interval Aggregate output and larger dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T21:56:01.950089Z",
     "start_time": "2020-05-18T21:55:05.400511Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing files ##\u001b[0m\n",
      "\u001b[32m\tReading input files header and checking consistancy between headers\u001b[0m\n",
      "\u001b[32m\tStarting asynchronous file parsing\u001b[0m\n",
      "\tProgress: 556M bytes [00:47, 11.7M bytes/s]                       \n",
      "\u001b[32m\tAdjust pvalues\u001b[0m\n",
      "\u001b[32m\tWriting output file\u001b[0m\n",
      "\tProgress: 100%|██████████| 266k/266k [00:08<00:00, 31.1k sites/s] \n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tInsufficient effect size: 201,365\u001b[0m\n",
      "\u001b[32m\t\tInsufficient samples: 54,136\u001b[0m\n",
      "\u001b[32m\t\tValid: 10,784\u001b[0m\n",
      "\u001b[32m\t\tNon-significant pvalue: 7,252\u001b[0m\n",
      "\u001b[32m\t\tSignificant pvalue: 3,532\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome start end   n_samples pvalue adj_pvalue neg_med pos_med ambiguous_med unique_cpg_pos labels                                   med_llr_list raw_llr_list raw_pos_list comment                  \n",
      "1          1657  1963  1         nan    nan        0       1       0             0              [\"7-2_F2\"]                               []           []           []           Insufficient samples     \n",
      "1          15653 15966 12        nan    nan        0       12      0             0              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[]           []           []           Insufficient effect size \n",
      "1          17092 17597 12        nan    nan        0       12      0             0              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[]           []           []           Insufficient effect size \n",
      "1          18071 18621 12        nan    nan        0       12      0             0              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[]           []           []           Insufficient effect size \n",
      "1          20376 21340 11        nan    nan        0       11      0             0              [\"117-2_C4\",\"131-1_F4\",\"134-1_H4\",\"13...[]           []           []           Insufficient effect size \n",
      "1          21578 21938 12        nan    nan        0       12      0             0              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[]           []           []           Insufficient effect size \n",
      "1          27747 28080 12        nan    nan        0       10      2             0              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[]           []           []           Insufficient effect size \n",
      "1          28288 28629 12        nan    nan        0       12      0             0              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[]           []           []           Insufficient effect size \n",
      "1          31270 31833 12        nan    nan        0       12      0             0              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[]           []           []           Insufficient effect size \n",
      "\n",
      "track name=meth_comp itemRgb=On\n",
      "1\t1657\t1963\t.\t0\t.\t1657\t1963\t230,230,230\n",
      "1\t15653\t15966\t.\t0\t.\t15653\t15966\t230,230,230\n",
      "1\t17092\t17597\t.\t0\t.\t17092\t17597\t230,230,230\n",
      "1\t18071\t18621\t.\t0\t.\t18071\t18621\t230,230,230\n",
      "1\t20376\t21340\t.\t0\t.\t20376\t21340\t230,230,230\n",
      "1\t21578\t21938\t.\t0\t.\t21578\t21938\t230,230,230\n",
      "1\t27747\t28080\t.\t0\t.\t27747\t28080\t230,230,230\n",
      "1\t28288\t28629\t.\t0\t.\t28288\t28629\t230,230,230\n",
      "1\t31270\t31833\t.\t0\t.\t31270\t31833\t230,230,230\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Generate list of file paths and sample ids from source directory\n",
    "from glob import glob\n",
    "import os\n",
    "fn_list = sorted(glob(\"./data/medaka_CGI_*\"))\n",
    "id_list = [os.path.split(fn)[-1].strip(\"medaka_CGI_\").strip(\".tsv.gz\") for fn in fn_list]\n",
    "\n",
    "Meth_Comp (\n",
    "    aggregate_fn_list=fn_list,\n",
    "    ref_fasta_fn=\"./data/medaka.fa\",\n",
    "    output_bed_fn=\"./results/Medaka_CGI_meth_comp.bed\",\n",
    "    output_tsv_fn=\"./results/Medaka_CGI_meth_comp.tsv.gz\",\n",
    "    sample_id_list = id_list,\n",
    "    max_missing = 1,\n",
    "    min_diff_llr = 1,\n",
    "    progress=True)\n",
    "\n",
    "head(\"./results/Medaka_CGI_meth_comp.tsv.gz\", max_char_col=40)\n",
    "head(\"./results/Medaka_CGI_meth_comp.bed\", max_char_col=40)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python [conda env:pycoMeth]",
   "language": "python",
   "name": "conda-env-pycoMeth-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
