{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Meth_Comp CLI usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Activate virtual environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T21:56:42.413426Z",
     "start_time": "2020-05-18T21:56:42.145284Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(pycoMeth) (pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Activate Conda env \n",
    "conda activate pycoMeth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T21:56:47.018382Z",
     "start_time": "2020-05-18T21:56:45.613687Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: pycoMeth Meth_Comp [-h] -i [AGGREGATE_FN_LIST [AGGREGATE_FN_LIST ...]]\n",
      "                          -f REF_FASTA_FN [-b OUTPUT_BED_FN]\n",
      "                          [-t OUTPUT_TSV_FN] [-m MAX_MISSING]\n",
      "                          [-l MIN_DIFF_LLR]\n",
      "                          [-s [SAMPLE_ID_LIST [SAMPLE_ID_LIST ...]]]\n",
      "                          [--pvalue_adj_method PVALUE_ADJ_METHOD]\n",
      "                          [--pvalue_threshold PVALUE_THRESHOLD]\n",
      "                          [--only_tested_sites] [-v] [-q] [-p]\n",
      "\n",
      "Compare methylation values for each CpG positions or intervals between n\n",
      "samples and perform a statistical test to evaluate if the positions are\n",
      "significantly different. For 2 samples a Mann_Withney test is performed\n",
      "otherwise multiples samples are compared with a Kruskal Wallis test. pValues\n",
      "are adjusted for multiple tests using the Benjamini & Hochberg procedure for\n",
      "controlling the false discovery rate.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input/Output options:\n",
      "  -i [AGGREGATE_FN_LIST [AGGREGATE_FN_LIST ...]], --aggregate_fn_list [AGGREGATE_FN_LIST [AGGREGATE_FN_LIST ...]]\n",
      "                        A list of output tsv files corresponding to several\n",
      "                        samples to compare generated by either CpG_Aggregate\n",
      "                        or Interval_Aggregate. (can be gzipped) (required)\n",
      "                        [str]\n",
      "  -f REF_FASTA_FN, --ref_fasta_fn REF_FASTA_FN\n",
      "                        Reference file used for alignment in Fasta format\n",
      "                        (ideally already indexed with samtools faidx)\n",
      "                        (required) [str]\n",
      "  -b OUTPUT_BED_FN, --output_bed_fn OUTPUT_BED_FN\n",
      "                        Path to write a summary result file in BED format (At\n",
      "                        least 1 output file is required) (can be gzipped)\n",
      "                        (default: None) [str]\n",
      "  -t OUTPUT_TSV_FN, --output_tsv_fn OUTPUT_TSV_FN\n",
      "                        Path to write an more extensive result report in TSV\n",
      "                        format (At least 1 output file is required) (can be\n",
      "                        gzipped) (default: None) [str]\n",
      "\n",
      "Misc options:\n",
      "  -m MAX_MISSING, --max_missing MAX_MISSING\n",
      "                        Max number of missing samples to perform the test\n",
      "                        (default: 0) [int]\n",
      "  -l MIN_DIFF_LLR, --min_diff_llr MIN_DIFF_LLR\n",
      "                        Minimal llr boundary for negative and positive median\n",
      "                        llr. The test if only performed if at least one sample\n",
      "                        has a median llr above (methylated) and 1 sample has a\n",
      "                        median llr below (unmethylated) (default: 2) [float]\n",
      "  -s [SAMPLE_ID_LIST [SAMPLE_ID_LIST ...]], --sample_id_list [SAMPLE_ID_LIST [SAMPLE_ID_LIST ...]]\n",
      "                        list of sample ids to annotate results in tsv file\n",
      "                        (default: None) [str]\n",
      "  --pvalue_adj_method PVALUE_ADJ_METHOD\n",
      "                        Method to use for pValue multiple test adjustment\n",
      "                        (default: fdr_bh) [str]\n",
      "  --pvalue_threshold PVALUE_THRESHOLD\n",
      "                        Alpha parameter (family-wise error rate) for pValue\n",
      "                        adjustment (default: 0.01) [float]\n",
      "  --only_tested_sites   Do not include sites that were not tested because of\n",
      "                        insufficient samples or effect size in the report\n",
      "                        (default: False) [None]\n",
      "\n",
      "Verbosity options:\n",
      "  -v, --verbose         Increase verbosity\n",
      "  -q, --quiet           Reduce verbosity\n",
      "  -p, --progress        Display a progress bar\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Meth_Comp --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with CpG Aggregate output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T21:57:27.553967Z",
     "start_time": "2020-05-18T21:57:16.806227Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing files ##\u001b[0m\n",
      "\u001b[32m\tReading input files header and checking consistancy between headers\u001b[0m\n",
      "\u001b[32m\tStarting asynchronous file parsing\u001b[0m\n",
      "\tProgress: 37.1M bytes [00:08, 4.31M bytes/s]                                   \n",
      "\u001b[32m\tAdjust pvalues\u001b[0m\n",
      "\u001b[32m\tWriting output file\u001b[0m\n",
      "\tProgress: 100%|██████████████████████| 3.28k/3.28k [00:00<00:00, 21.8k sites/s]\n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tInsufficient samples: 211,400\u001b[0m\n",
      "\u001b[32m\t\tInsufficient effect size: 30,132\u001b[0m\n",
      "\u001b[32m\t\tValid: 3,283\u001b[0m\n",
      "\u001b[32m\t\tNon-significant pvalue: 3,283\u001b[0m\n",
      "(pycoMeth) (pycoMeth) track name=meth_comp itemRgb=On\n",
      "I\t109610\t109611\t.\t0\t.\t109610\t109611\t230,230,230\n",
      "I\t109667\t109668\t.\t0\t.\t109667\t109668\t230,230,230\n",
      "I\t110046\t110047\t.\t0\t.\t110046\t110047\t230,230,230\n",
      "I\t110157\t110158\t.\t0\t.\t110157\t110158\t230,230,230\n",
      "I\t110226\t110227\t.\t0\t.\t110226\t110227\t230,230,230\n",
      "I\t110246\t110247\t.\t0\t.\t110246\t110247\t230,230,230\n",
      "I\t110689\t110690\t.\t0\t.\t110689\t110690\t230,230,230\n",
      "I\t110925\t110926\t.\t0\t.\t110925\t110926\t230,230,230\n",
      "I\t111021\t111022\t.\t0\t.\t111021\t111022\t230,230,230\n",
      "(pycoMeth) chromosome\tstart\tend\tn_samples\tpvalue\tadj_pvalue\tneg_med\tpos_med\tambiguous_med\tlabels\tmed_llr_list\traw_llr_list\tcomment\n",
      "I\t109610\t109611\t4\t0.5355384957105982\t0.637278314240832\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[1.06,-3.0,-3.285,-3.575]\t[[2.83,-0.71],[-4.85,1.83,-3.0],[0.87,-7.44],[0.13,-4.22,-2.93,-6.8]]\tNon-significant pvalue\n",
      "I\t109667\t109668\t4\t0.6210918363528521\t0.6997407339555297\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[0.555,-1.45,-4.96,-2.57]\t[[5.93,-4.82],[-1.45,-1.28,-5.4],[-5.79,-4.13],[-5.81,-1.87,-0.82,-2.57,-5.08]]\tNon-significant pvalue\n",
      "I\t110046\t110047\t4\t0.4846302240896865\t0.6031859226921511\t2\t1\t1\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[0.3,-1.33,-3.355,1.32]\t[[-2.0,2.6],[-1.94,-1.33,-1.2],[-1.73,-4.98],[3.33,1.8,-1.03,-5.67,1.32]]\tNon-significant pvalue\n",
      "I\t110157\t110158\t4\t0.10079374072306878\t0.4770356012343023\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-0.535,1.14,-3.28,-2.96]\t[[1.03,-2.1],[1.14,3.45,-0.49],[-2.79,-3.77],[-3.61,-0.77,-2.96,-0.32,-7.67]]\tNon-significant pvalue\n",
      "I\t110226\t110227\t4\t0.2562512982158426\t0.48938810639848035\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[1.025,-1.21,-0.875,-2.85]\t[[1.13,0.92],[0.46,-1.45,-1.21],[-1.62,-0.13],[-6.69,2.96,-4.64,-1.5,-2.85]]\tNon-significant pvalue\n",
      "I\t110246\t110247\t4\t0.16789646023633203\t0.4770356012343023\t2\t1\t1\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-2.63,0.57,-2.05,0.2]\t[[-3.84,-1.42],[-1.04,0.91,0.57],[-3.46,-0.64],[-3.17,0.17,0.94,0.2,1.83]]\tNon-significant pvalue\n",
      "I\t110689\t110690\t4\t0.15407920997458144\t0.4770356012343023\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-1.885,0.5,-6.53,-2.37]\t[[-1.89,-1.88],[1.32,-0.32],[-2.53,-10.53],[-1.76,0.49,-5.91,-2.37,-4.05]]\tNon-significant pvalue\n",
      "I\t110925\t110926\t4\t0.45883970397306584\t0.5920104435099945\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-3.06,-5.365,-1.785,0.61]\t[[-3.06,-4.18,2.33],[-10.58,-0.15],[-5.51,1.94],[1.84,-1.45,0.61,2.65,0.35]]\tNon-significant pvalue\n",
      "I\t111021\t111022\t4\t0.10926389869102016\t0.4770356012343023\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-7.8,-4.85,-4.94,0.51]\t[[-7.8,-3.83,-8.36],[-3.64,-6.06],[-4.19,-5.69],[2.71,1.03,0.32,-5.6,0.51]]\tNon-significant pvalue\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Meth_Comp \\\n",
    "    -i  \"./data/Yeast_CpG_1.tsv.gz\" \"./data/Yeast_CpG_2.tsv.gz\" \"./data/Yeast_CpG_3.tsv.gz\" \"./data/Yeast_CpG_4.tsv.gz\" \\\n",
    "    -f \"./data/yeast.fa\" \\\n",
    "    -b \"./results/CLI_Yeast_CpG_meth_comp.bed\" \\\n",
    "    -t \"./results/CLI_Yeast_CpG_meth_comp.tsv\" \\\n",
    "    -s S1 S2 S3 S4 \\\n",
    "    -l 0.5 \\\n",
    "    --pvalue_threshold 0.05 \\\n",
    "    --only_tested_sites \\\n",
    "    --progress\n",
    "\n",
    "head ./results/CLI_Yeast_CpG_meth_comp.bed\n",
    "head ./results/CLI_Yeast_CpG_meth_comp.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with Interval Aggregate output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-05-18T21:58:21.451923Z",
     "start_time": "2020-05-18T21:57:32.656444Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing files ##\u001b[0m\n",
      "\u001b[32m\tReading input files header and checking consistancy between headers\u001b[0m\n",
      "\u001b[32m\tStarting asynchronous file parsing\u001b[0m\n",
      "\tProgress: 556M bytes [00:40, 13.6M bytes/s]                                    \n",
      "\u001b[32m\tAdjust pvalues\u001b[0m\n",
      "\u001b[32m\tWriting output file\u001b[0m\n",
      "\tProgress: 100%|████████████████████████| 266k/266k [00:05<00:00, 45.8k sites/s]\n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tInsufficient effect size: 156,368\u001b[0m\n",
      "\u001b[32m\t\tInsufficient samples: 108,385\u001b[0m\n",
      "\u001b[32m\t\tValid: 1,532\u001b[0m\n",
      "\u001b[32m\t\tSignificant pvalue: 1,106\u001b[0m\n",
      "\u001b[32m\t\tNon-significant pvalue: 426\u001b[0m\n",
      "(pycoMeth) (pycoMeth) track name=meth_comp itemRgb=On\n",
      "1\t1657\t1963\t.\t0\t.\t1657\t1963\t230,230,230\n",
      "1\t15653\t15966\t.\t0\t.\t15653\t15966\t230,230,230\n",
      "1\t17092\t17597\t.\t0\t.\t17092\t17597\t230,230,230\n",
      "1\t18071\t18621\t.\t0\t.\t18071\t18621\t230,230,230\n",
      "1\t20376\t21340\t.\t0\t.\t20376\t21340\t230,230,230\n",
      "1\t21578\t21938\t.\t0\t.\t21578\t21938\t230,230,230\n",
      "1\t27747\t28080\t.\t0\t.\t27747\t28080\t230,230,230\n",
      "1\t28288\t28629\t.\t0\t.\t28288\t28629\t230,230,230\n",
      "1\t31270\t31833\t.\t0\t.\t31270\t31833\t230,230,230\n",
      "(pycoMeth) chromosome\tstart\tend\tn_samples\tpvalue\tadj_pvalue\tneg_med\tpos_med\tambiguous_med\tunique_cpg_pos\tlabels\tmed_llr_list\traw_llr_list\traw_pos_list\tcomment\n",
      "1\t1657\t1963\t1\tnan\tnan\t0\t1\t0\t0\t[9]\t[]\t[]\t[]\tInsufficient samples\n",
      "1\t15653\t15966\t12\tnan\tnan\t0\t12\t0\t0\t[0,1,2,3,4,5,6,7,8,9,10,11]\t[]\t[]\t[]\tInsufficient effect size\n",
      "1\t17092\t17597\t12\tnan\tnan\t0\t6\t6\t0\t[0,1,2,3,4,5,6,7,8,9,10,11]\t[]\t[]\t[]\tInsufficient effect size\n",
      "1\t18071\t18621\t12\tnan\tnan\t0\t12\t0\t0\t[0,1,2,3,4,5,6,7,8,9,10,11]\t[]\t[]\t[]\tInsufficient effect size\n",
      "1\t20376\t21340\t11\tnan\tnan\t0\t11\t0\t0\t[1,2,3,4,5,6,7,8,9,10,11]\t[]\t[]\t[]\tInsufficient samples\n",
      "1\t21578\t21938\t12\tnan\tnan\t0\t10\t2\t0\t[0,1,2,3,4,5,6,7,8,9,10,11]\t[]\t[]\t[]\tInsufficient effect size\n",
      "1\t27747\t28080\t12\tnan\tnan\t0\t6\t6\t0\t[0,1,2,3,4,5,6,7,8,9,10,11]\t[]\t[]\t[]\tInsufficient effect size\n",
      "1\t28288\t28629\t12\tnan\tnan\t0\t11\t1\t0\t[0,1,2,3,4,5,6,7,8,9,10,11]\t[]\t[]\t[]\tInsufficient effect size\n",
      "1\t31270\t31833\t12\tnan\tnan\t0\t8\t4\t0\t[0,1,2,3,4,5,6,7,8,9,10,11]\t[]\t[]\t[]\tInsufficient effect size\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Meth_Comp \\\n",
    "    -i ./data/medaka_CGI_* \\\n",
    "    -f \"./data/medaka.fa\" \\\n",
    "    -b \"./results/CLI_Medaka_CGI_meth_comp.bed\" \\\n",
    "    -t \"./results/CLI_Medaka_CGI_meth_comp.tsv\" \\\n",
    "    --progress\n",
    "\n",
    "head ./results/CLI_Medaka_CGI_meth_comp.bed\n",
    "head ./results/CLI_Medaka_CGI_meth_comp.tsv"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
